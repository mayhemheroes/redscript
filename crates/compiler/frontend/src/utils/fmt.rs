use core::fmt;
use std::cell::RefCell;

#[derive(Debug, Clone)]
pub struct DisplayFn<F>(F);

impl<F> DisplayFn<F> {
    pub fn new(f: F) -> Self {
        Self(f)
    }
}

impl<F> fmt::Display for DisplayFn<F>
where
    F: Fn(&mut fmt::Formatter<'_>) -> fmt::Result,
{
    #[inline]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0(f)
    }
}

pub fn lowercase(thing: &str) -> impl fmt::Display + Clone + use<'_> {
    DisplayFn::new(move |f: &mut fmt::Formatter<'_>| {
        thing
            .chars()
            .try_for_each(|c| write!(f, "{}", c.to_lowercase()))
    })
}

pub fn surrounded_by<'a, A>(
    thing: A,
    left: &'a str,
    right: &'a str,
) -> impl fmt::Display + Clone + use<'a, A>
where
    A: fmt::Display + Clone,
{
    DisplayFn::new(move |f: &mut fmt::Formatter<'_>| write!(f, "{left}{thing}{right}"))
}

pub fn sep_by<T, I>(iter: I, sep: &str) -> impl fmt::Display + use<'_, T, I>
where
    I: IntoIterator<Item = T>,
    I::Item: fmt::Display,
{
    let iter = RefCell::new(iter.into_iter());
    DisplayFn::new(move |f: &mut fmt::Formatter<'_>| {
        if let Some(first) = iter.borrow_mut().next() {
            write!(f, "{first}")?;
        }
        iter.borrow_mut()
            .try_for_each(|item| write!(f, "{sep}{item}"))?;
        Ok(())
    })
}
