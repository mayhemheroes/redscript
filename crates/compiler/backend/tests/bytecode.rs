use std::cell::RefCell;
use std::fmt;

use indexmap::IndexSet;
use redscript_compiler_api::{Diagnostics, SourceMap, SourceMapExt, TypeInterner};
use redscript_compiler_backend::{CompilationInputs, Monomorphizer};
use redscript_compiler_frontend::infer_from_sources;
use redscript_io::{
    CNameIndex, Class, ClassFlags, ClassIndex, EnumIndex, EnumValueIndex, FieldIndex,
    FunctionIndex, IndexedDefinition, Instr, LocalIndex, ParameterIndex, ResourceIndex,
    ScriptBundle, StringIndex, TweakDbIndex, TypeIndex, Visibility,
};

#[test]
fn bytecode() {
    insta::glob!("data/*.reds", |path| {
        let mut sources = SourceMap::from_files(&[path]).unwrap();
        sources.populate_boot_lib();

        let interner = TypeInterner::default();
        let mut bundle = ScriptBundle::default();

        let name = bundle.cnames_mut().add("IScriptable");
        bundle.define(Class::new(name, Visibility::Public, ClassFlags::default()));

        let (symbols, mappings) = CompilationInputs::load(&bundle, &interner)
            .unwrap()
            .into_inner();
        let (unit, symbols, diagnostics) = infer_from_sources(&sources, &interner, symbols);
        assert_eq!(diagnostics.len(), 0, "{}", Diagnostics::from(diagnostics));

        let bundle_len = bundle.definitions().len();

        Monomorphizer::from(mappings)
            .monomorphize(&unit, &symbols, &mut bundle)
            .unwrap();
        insta::assert_snapshot!(BytecodePrinter::new(&bundle, bundle_len));
    })
}

#[derive(Debug)]
struct BytecodePrinter<'a> {
    bundle: &'a ScriptBundle<'a>,
    offset: usize,
    indices: RefCell<Indices>,
}

impl<'a> BytecodePrinter<'a> {
    fn new(bundle: &'a ScriptBundle<'a>, offset: usize) -> Self {
        Self {
            bundle,
            offset,
            indices: RefCell::new(Indices::default()),
        }
    }
}

impl fmt::Display for BytecodePrinter<'_> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut indices = self.indices.borrow_mut();

        for def in self.bundle.indexed_definitions().skip(self.offset) {
            let IndexedDefinition::Function(i, func) = def else {
                continue;
            };

            write!(f, "// ")?;
            if let Some(class) = func.class() {
                write!(f, "{}::", self.bundle[self.bundle[class].name()])?;
            }
            writeln!(f, "{} ({})", self.bundle[func.name()], indices.function(i))?;

            for instr in func.body().code_iter() {
                match instr.unwrap() {
                    Instr::CNameConst(idx) => {
                        writeln!(f, "cname.const {}", indices.name(idx))?;
                    }
                    Instr::EnumConst { enum_, value } => {
                        writeln!(
                            f,
                            "enum.const {} {}",
                            indices.enum_(enum_),
                            indices.enum_value(value)
                        )?;
                    }
                    Instr::StringConst(idx) => {
                        writeln!(f, "string.const {}", indices.string(idx))?;
                    }
                    Instr::TweakDbIdConst(idx) => {
                        writeln!(f, "tweakdb.const {}", indices.tweakdb(idx))?;
                    }
                    Instr::ResourceConst(idx) => {
                        writeln!(f, "resource.const {}", indices.resource(idx))?;
                    }
                    Instr::Local(idx) => {
                        writeln!(f, "local {}", indices.local(idx))?;
                    }
                    Instr::Param(idx) => {
                        writeln!(f, "param {}", indices.param(idx))?;
                    }
                    Instr::ObjectField(idx) => {
                        writeln!(f, "object.field {}", indices.field(idx))?;
                    }
                    Instr::Construct { arg_count, class } => {
                        writeln!(f, "struct.new {} {}", arg_count, indices.class(class))?;
                    }
                    Instr::InvokeStatic {
                        exit,
                        line,
                        function,
                        flags,
                    } => {
                        writeln!(
                            f,
                            "invoke.static j{} l{} f{} {}",
                            exit.target(),
                            line,
                            indices.function(function),
                            flags
                        )?;
                    }
                    Instr::InvokeVirtual {
                        exit,
                        line,
                        function,
                        flags,
                    } => {
                        writeln!(
                            f,
                            "invoke.virtual j{} l{} f{} {}",
                            exit.target(),
                            line,
                            indices.name(function),
                            flags
                        )?;
                    }
                    Instr::StructField(idx) => {
                        writeln!(f, "struct.field {}", indices.field(idx))?;
                    }
                    Instr::Equals(idx) => {
                        writeln!(f, "eq {}", indices.type_(idx))?;
                    }
                    Instr::RefStringEqualsString(idx) => {
                        writeln!(f, "refstr.eq {}", indices.type_(idx))?;
                    }
                    Instr::StringEqualsRefString(idx) => {
                        writeln!(f, "str.eq {}", indices.type_(idx))?;
                    }
                    Instr::NotEquals(idx) => {
                        writeln!(f, "neq {}", indices.type_(idx))?;
                    }
                    Instr::RefStringNotEqualsString(idx) => {
                        writeln!(f, "refstr.neq {}", indices.type_(idx))?;
                    }
                    Instr::StringNotEqualsRefString(idx) => {
                        writeln!(f, "str.neq {}", indices.type_(idx))?;
                    }
                    Instr::New(idx) => {
                        writeln!(f, "object.new {}", indices.class(idx))?;
                    }
                    Instr::ArrayClear(idx) => writeln!(f, "array.clear {}", indices.type_(idx))?,
                    Instr::ArraySize(idx) => writeln!(f, "array.size {}", indices.type_(idx))?,
                    Instr::ArrayResize(idx) => writeln!(f, "array.resize {}", indices.type_(idx))?,
                    Instr::ArrayFindFirst(idx) => {
                        writeln!(f, "array.find_first {}", indices.type_(idx))?
                    }
                    Instr::ArrayFindFirstFast(idx) => {
                        writeln!(f, "array.find_first_fast {}", indices.type_(idx))?
                    }
                    Instr::ArrayFindLast(idx) => {
                        writeln!(f, "array.find_last {}", indices.type_(idx))?
                    }
                    Instr::ArrayFindLastFast(idx) => {
                        writeln!(f, "array.find_last_fast {}", indices.type_(idx))?
                    }
                    Instr::ArrayContains(idx) => {
                        writeln!(f, "array.contains {}", indices.type_(idx))?
                    }
                    Instr::ArrayContainsFast(idx) => {
                        writeln!(f, "array.contains_fast {}", indices.type_(idx))?
                    }
                    Instr::ArrayCount(idx) => writeln!(f, "array.count {}", indices.type_(idx))?,
                    Instr::ArrayCountFast(idx) => {
                        writeln!(f, "array.count_fast {}", indices.type_(idx))?
                    }
                    Instr::ArrayPush(idx) => writeln!(f, "array.push {}", indices.type_(idx))?,
                    Instr::ArrayPop(idx) => writeln!(f, "array.pop {}", indices.type_(idx))?,
                    Instr::ArrayInsert(idx) => writeln!(f, "array.insert {}", indices.type_(idx))?,
                    Instr::ArrayRemove(idx) => writeln!(f, "array.remove {}", indices.type_(idx))?,
                    Instr::ArrayRemoveFast(idx) => {
                        writeln!(f, "array.remove_fast {}", indices.type_(idx))?
                    }
                    Instr::ArrayGrow(idx) => writeln!(f, "array.grow {}", indices.type_(idx))?,
                    Instr::ArrayErase(idx) => writeln!(f, "array.erase {}", indices.type_(idx))?,
                    Instr::ArrayEraseFast(idx) => {
                        writeln!(f, "array.erase_fast {}", indices.type_(idx))?
                    }
                    Instr::ArrayLast(idx) => writeln!(f, "array.last {}", indices.type_(idx))?,
                    Instr::ArrayElement(idx) => {
                        writeln!(f, "array.element {}", indices.type_(idx))?
                    }
                    Instr::ArraySort(idx) => writeln!(f, "array.sort {}", indices.type_(idx))?,
                    Instr::ArraySortByPredicate(idx) => {
                        writeln!(f, "array.sort_by {}", indices.type_(idx))?
                    }
                    Instr::StaticArraySize(idx) => {
                        writeln!(f, "static_array.size {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayFindFirst(idx) => {
                        writeln!(f, "static_array.find_first {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayFindFirstFast(idx) => {
                        writeln!(f, "static_array.find_first_fast {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayFindLast(idx) => {
                        writeln!(f, "static_array.find_last {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayFindLastFast(idx) => {
                        writeln!(f, "static_array.find_last_fast {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayContains(idx) => {
                        writeln!(f, "static_array.contains {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayContainsFast(idx) => {
                        writeln!(f, "static_array.contains_fast {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayCount(idx) => {
                        writeln!(f, "static_array.count {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayCountFast(idx) => {
                        writeln!(f, "static_array.count_fast {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayLast(idx) => {
                        writeln!(f, "static_array.last {}", indices.type_(idx))?
                    }
                    Instr::StaticArrayElement(idx) => {
                        writeln!(f, "static_array.element {}", indices.type_(idx))?
                    }
                    Instr::EnumToI32 { enum_type, size } => {
                        writeln!(f, "enum.to_int {} {}", indices.type_(enum_type), size)?
                    }
                    Instr::I32ToEnum { enum_type, size } => {
                        writeln!(f, "enum.from_int {} {}", indices.type_(enum_type), size)?
                    }
                    Instr::DynamicCast { class, is_weak } if is_weak => {
                        writeln!(f, "wref.dyncast {}", indices.class(class))?
                    }
                    Instr::DynamicCast { class, .. } => {
                        writeln!(f, "ref.dyncast {}", indices.class(class))?
                    }
                    Instr::ToString(idx) => {
                        writeln!(f, "to_string {}", indices.type_(idx))?;
                    }
                    Instr::ToVariant(idx) => {
                        writeln!(f, "variant.new {}", indices.type_(idx))?;
                    }
                    Instr::FromVariant(idx) => {
                        writeln!(f, "variant.extract {}", indices.type_(idx))?;
                    }
                    Instr::AsRef(idx) => {
                        writeln!(f, "as_ref {}", indices.type_(idx))?;
                    }
                    Instr::Deref(idx) => {
                        writeln!(f, "deref {}", indices.type_(idx))?;
                    }
                    other => writeln!(f, "{}", other)?,
                }
            }

            writeln!(f)?;
        }

        Ok(())
    }
}

#[derive(Debug, Default)]
struct Indices {
    names: IndexSet<CNameIndex>,
    enums: IndexSet<EnumIndex>,
    enum_values: IndexSet<EnumValueIndex>,
    strings: IndexSet<StringIndex>,
    tweakdb_ids: IndexSet<TweakDbIndex>,
    resouces: IndexSet<ResourceIndex>,
    locals: IndexSet<LocalIndex>,
    params: IndexSet<ParameterIndex>,
    fields: IndexSet<FieldIndex>,
    classes: IndexSet<ClassIndex>,
    functions: IndexSet<FunctionIndex>,
    types: IndexSet<TypeIndex>,
}

impl Indices {
    fn name(&mut self, index: CNameIndex) -> usize {
        self.names.insert_full(index).0
    }

    fn enum_(&mut self, index: EnumIndex) -> usize {
        self.enums.insert_full(index).0
    }

    fn enum_value(&mut self, index: EnumValueIndex) -> usize {
        self.enum_values.insert_full(index).0
    }

    fn string(&mut self, index: StringIndex) -> usize {
        self.strings.insert_full(index).0
    }

    fn tweakdb(&mut self, index: TweakDbIndex) -> usize {
        self.tweakdb_ids.insert_full(index).0
    }

    fn resource(&mut self, index: ResourceIndex) -> usize {
        self.resouces.insert_full(index).0
    }

    fn local(&mut self, index: LocalIndex) -> usize {
        self.locals.insert_full(index).0
    }

    fn param(&mut self, index: ParameterIndex) -> usize {
        self.params.insert_full(index).0
    }

    fn field(&mut self, index: FieldIndex) -> usize {
        self.fields.insert_full(index).0
    }

    fn class(&mut self, index: ClassIndex) -> usize {
        self.classes.insert_full(index).0
    }

    fn function(&mut self, index: FunctionIndex) -> usize {
        self.functions.insert_full(index).0
    }

    fn type_(&mut self, index: TypeIndex) -> usize {
        self.types.insert_full(index).0
    }
}
