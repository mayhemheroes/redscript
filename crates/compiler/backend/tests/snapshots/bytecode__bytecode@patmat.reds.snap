---
source: crates/compiler/backend/tests/bytecode.rs
expression: "BytecodePrinter::new(&bundle, bundle_len)"
input_file: crates/compiler/backend/tests/data/patmat.reds
---
// Scriptable::IsA;CName (f0)

// TestArray;array<Int32> (f1)
switch 5
true.const
switch.label 248 39
invoke.static j39 l4 f2 0000000000000000
array.size 0
param 0
i32.const 3
param.end
assign
local 0
array.element 0
param 0
invoke.static j39 l4 f3 0000000000000000
array.size 0
param 0
i32.const 3
param.end
assign
local 1
array.element 0
param 0
invoke.static j39 l4 f3 0000000000000000
array.size 0
param 0
i32.const 2
param.end
assign
local 2
array.element 0
param 0
invoke.static j35 l4 f3 0000000000000000
array.size 0
param 0
i32.one
param.end
return
local 0
switch.label 109 39
invoke.static j39 l6 f2 0000000000000000
array.size 0
param 0
i32.const 2
param.end
assign
local 3
array.element 0
param 0
i32.zero
assign
local 4
array.element 0
param 0
i32.one
return
local 4
switch.label 76 35
invoke.static j35 l8 f4 0000000000000000
array.size 0
param 0
i32.one
param.end
assign
local 5
array.element 0
param 0
i32.zero
return
local 5
switch.default
return
i32.zero
assign
local 6
param 0
assign
local 7
param 0
assign
local 8
param 0
jump.if_not 249
invoke.static j39 l14 f2 0000000000000000
array.size 0
local 6
i32.const 3
param.end
assign
local 9
array.element 0
local 6
invoke.static j39 l14 f3 0000000000000000
array.size 0
local 6
i32.const 3
param.end
assign
local 10
array.element 0
local 6
invoke.static j39 l14 f3 0000000000000000
array.size 0
local 6
i32.const 2
param.end
assign
local 11
array.element 0
local 6
invoke.static j35 l14 f3 0000000000000000
array.size 0
local 6
i32.one
param.end
return
local 9
jump 195
jump.if_not 110
invoke.static j39 l16 f2 0000000000000000
array.size 0
local 7
i32.const 2
param.end
assign
local 12
array.element 0
local 7
i32.zero
assign
local 13
array.element 0
local 7
i32.one
return
local 13
jump 82
jump.if_not 77
invoke.static j35 l18 f4 0000000000000000
array.size 0
local 8
i32.one
param.end
assign
local 14
array.element 0
local 8
i32.zero
return
local 14
jump 2
return
i32.zero

// TestNullable;Class (f5)
switch 5
true.const
switch.label 41 10
ref.to_bool
param 1
assign
local 15
param 1
return
local 15
switch.default
return
ref.null
assign
local 16
param 1
jump.if_not 42
ref.to_bool
local 16
assign
local 17
local 16
return
local 17
jump 2
return
ref.null

// TestNullableWref;Class (f6)
switch 5
true.const
switch.label 43 11
ref.to_bool
wref.to_ref
param 2
assign
local 18
param 2
return
wref.to_ref
local 18
switch.default
return
ref.null
assign
local 19
param 2
jump.if_not 44
ref.to_bool
wref.to_ref
local 19
assign
local 20
local 19
return
wref.to_ref
local 20
jump 2
return
ref.null

// TestDestructure;Class (f7)
switch 5
true.const
switch.label 80 37
ctx 34
param 3
invoke.virtual j25 l57 f0 0000000000000000
cname.const 1
param.end
assign
local 21
ctx 18
param 3
object.field 0
return
local 21
switch.label 90 37
ctx 34
param 3
invoke.virtual j25 l59 f0 0000000000000000
cname.const 2
param.end
assign
local 22
ctx 18
param 3
object.field 1
return
array.element 0
local 22
i32.zero
switch.default
return
i32.zero
assign
local 23
param 3
assign
local 24
param 3
jump.if_not 81
ctx 34
local 23
invoke.virtual j25 l65 f0 0000000000000000
cname.const 1
param.end
assign
local 25
ctx 18
local 23
object.field 0
return
local 25
jump 96
jump.if_not 91
ctx 34
local 24
invoke.virtual j25 l67 f0 0000000000000000
cname.const 2
param.end
assign
local 26
ctx 18
local 24
object.field 1
return
array.element 0
local 26
i32.zero
jump 2
return
i32.zero

// TestDestructureWref;Class (f8)
switch 5
true.const
switch.label 82 38
ctx 35
wref.to_ref
param 4
invoke.virtual j25 l76 f0 0000000000000000
cname.const 1
param.end
assign
local 27
ctx 19
wref.to_ref
param 4
object.field 0
return
local 27
switch.label 92 38
ctx 35
wref.to_ref
param 4
invoke.virtual j25 l78 f0 0000000000000000
cname.const 2
param.end
assign
local 28
ctx 19
wref.to_ref
param 4
object.field 1
return
array.element 0
local 28
i32.zero
switch.default
return
i32.zero
assign
local 29
param 4
assign
local 30
param 4
jump.if_not 83
ctx 35
wref.to_ref
local 29
invoke.virtual j25 l84 f0 0000000000000000
cname.const 1
param.end
assign
local 31
ctx 19
wref.to_ref
local 29
object.field 0
return
local 31
jump 98
jump.if_not 93
ctx 35
wref.to_ref
local 30
invoke.virtual j25 l86 f0 0000000000000000
cname.const 2
param.end
assign
local 32
ctx 19
wref.to_ref
local 30
object.field 1
return
array.element 0
local 32
i32.zero
jump 2
return
i32.zero

// TestCast;Class (f9)
switch 5
true.const
switch.label 80 37
ctx 34
param 5
invoke.virtual j25 l95 f0 0000000000000000
cname.const 1
param.end
assign
local 33
param 5
return
ctx 18
local 33
object.field 0
switch.label 90 37
ctx 34
param 5
invoke.virtual j25 l97 f0 0000000000000000
cname.const 2
param.end
assign
local 34
param 5
return
array.element 0
ctx 18
local 34
object.field 1
i32.zero
switch.default
return
i32.zero
assign
local 35
param 5
assign
local 36
param 5
jump.if_not 81
ctx 34
local 35
invoke.virtual j25 l103 f0 0000000000000000
cname.const 1
param.end
assign
local 37
local 35
return
ctx 18
local 37
object.field 0
jump 96
jump.if_not 91
ctx 34
local 36
invoke.virtual j25 l105 f0 0000000000000000
cname.const 2
param.end
assign
local 38
local 36
return
array.element 0
ctx 18
local 38
object.field 1
i32.zero
jump 2
return
i32.zero

// TestWhile;Class (f10)
assign
local 39
param 6
jump.if_not 157
true.const
assign
local 40
local 39
jump.if_not 128
invoke.static j75 l114 f11 0000000000000000
ctx 34
local 40
invoke.virtual j25 l114 f0 0000000000000000
cname.const 3
param.end
ref.to_bool
ctx 18
local 40
object.field 2
param.end
assign
local 41
ctx 18
local 40
object.field 2
assign
local 39
local 41
jump 3
jump 3
jump -160
return
local 39

// TestWhileArray;array<Int32> (f12)
jump.if_not 148
true.const
assign
local 42
param 7
jump.if_not 119
invoke.static j35 l121 f2 0000000000000000
array.size 0
local 42
i32.one
param.end
assign
local 43
array.element 0
local 42
invoke.static j35 l121 f3 0000000000000000
array.size 0
local 42
i32.one
param.end
array.pop 0
param 7
jump 3
jump 3
jump -151

// OperatorGreaterEqual (f2)

// OperatorSubtract (f3)

// OperatorEqual (f4)

// OperatorLogicAnd (f11)
